lib/int64.js

```diff
@@ -0,0 +1,335 @@
+'use strict';
+
+const INT32_MAX = 0x7FFFFFFF;
+const UINT32_MAX = 0xFFFFFFFF;
+
+class Int64 {
+  constructor(value) {
+    this.value = new Uint32Array(2);
+    if (value instanceof Int64) {
+      this.value[0] = value.value[0];
+      this.value[1] = value.value[1];
+      return;
+    }
+    const numValue = Number(value);
+    if (Number.isNaN(numValue)) return;
+    if (numValue >= 0 && numValue <= INT32_MAX) {
+      this.value[0] = numValue;
+    } else if (numValue < 0 && numValue >= -INT32_MAX) {
+      this.value[0] = numValue;
+      this.value[1] = UINT32_MAX;
+    } else if (numValue >= 0 && numValue <= Number.MAX_SAFE_INTEGER) {
+      this.value[0] = numValue % (UINT32_MAX + 1);
+      this.value[1] = Math.floor(numValue / (UINT32_MAX + 1));
+    } else if (numValue < 0 && numValue >= Number.MIN_SAFE_INTEGER) {
+      this.value[0] = -numValue % (UINT32_MAX + 1);
+      this.value[1] = Math.floor(-numValue / (UINT32_MAX + 1));
+      this.value = Int64._conversion(this.value);
+    } else if (typeof value === 'string') {
+      let res = new Int64();
+      if (value.startsWith('0x')) {
+        value = value.slice(2);
+        res.value[0] = Number('0x' + value.slice(value.length - 8));
+        res.value[1] = Number('0x' + value.slice(0, value.length - 8));
+      } else if (value.startsWith('0o')) {
+        value = value.slice(2);
+        const val0 = Number('0o' + value.slice(value.length - 11));
+        const val1 = Number('0o' + value.slice(0, value.length - 11));
+        res.value[0] = val0;
+        res.value[1] = (val1 << 1) | ((val0 >> 31) & 1);
+      } else if (value.startsWith('0b')) {
+        value = value.slice(2);
+        res.value[0] = Number('0b' + value.slice(value.length - 32));
+        res.value[1] = Number('0b' + value.slice(0, value.length - 32));
+      } else {
+        const sign = value.startsWith('-');
+        if (sign) value = value.slice(1);
+        const intRadix = new Int64(10);
+        for (let i = 0; i < value.length; i++) {
+          const intDigit = new Int64(value[i]);
+          res = Int64.mult(res, intRadix).add(intDigit);
+        }
+        if (sign) res.value = Int64._conversion(res.value);
+      }
+      this.value[0] = res.value[0];
+      this.value[1] = res.value[1];
+    }
+  }
+
+  toInt32() {
+    return this.value[0];
+  }
+
+  static _conversion(value) {
+    const result = new Uint32Array(value);
+    let pos = 0;
+    if (result[0] === 0) {
+      while (pos < 32) {
+        if ((result[1] >>> pos) & 1) break;
+        pos++;
+      }
+      const mask = (1 << (pos + 1)) - 1;
```

Comment:
This will potentially shift by 32 (max pos in 31 and then +1) which will result in `1` and therefore `(1 << (pos + 1)) - 1` will be `0`, is that intentional?

Reply:
This also seems to lead to a bug when handling value mentioned in the https://github.com/metarhia/common/pull/203#discussion_r227176378.